experimental/LieAlgebras/src/CartanMatrix.jl

```diff
@@ -128,17 +144,280 @@ function is_cartan_matrix(mat::ZZMatrix; generalized::Bool=true)
+
+# Example
+```jldoctest
+julia> cartan_symmetrizer(cartan_matrix(:B, 2))
+2-element Vector{ZZRingElem}:
+ 2
+ 1
+"""
+function cartan_symmetrizer(gcm::ZZMatrix; check::Bool=true)
+  @req !check || is_cartan_matrix(gcm) "requires a generalized Cartan matrix"
+  rk = nrows(gcm)
+  diag = ones(ZZRingElem, rk)
 
-      # if we only want a generalized Cartan matrix, we are done with this pair of entries
-      if generalized
+  # used for traversal
+  undone = trues(rk)
+  plan = zeros(Int, rk) # roots planned sorted asc grouped by component
+  head = 0
+  tail = 0
+
+  # we collect roots of the same length
+  # once we know if they are short or long we scale approriately
+  while any(undone)
+    if head == tail
+      head += 1
+      plan[head] = findfirst(undone)
+      undone[plan[head]] = false
+    end
+
+    i = plan[head]
+    for j in 1:rk
+      if i == j
         continue
       end
 
-      if !(mat[i, j] * mat[j, i] in 1:3)
-        return false
+      if undone[j] && !is_zero_entry(gcm, i, j)
+        head += 1
+        plan[head] = j
+        undone[j] = false
+
+        if diag[i] * gcm[i, j] == diag[j] * gcm[j, i]
+          continue
+        elseif gcm[i, j] == gcm[j, i]
+          diag[i] = lcm(diag[i], diag[j])
+          diag[j] = diag[i]
+          continue
+        end
+
+        if gcm[j, i] < -1
+          tail += 1
+          v = -gcm[j, i]
+          while tail < head
+            diag[plan[tail]] *= v
+            tail += 1
+          end
+        end
+        if gcm[i, j] < -1
+          diag[j] *= -gcm[i, j]
+          tail = head - 1
+        end
+      end
+    end
+  end
+
+  return diag
+end
+
+@doc raw"""
+    cartan_bilinear_form(gcm::ZZMatrix; check::Bool=true) -> ZZMatrix
+
+Returns the matrix of the symmetric bilinear form associated to the Cartan matrix from `cartan_matrix_symmetrizer`.
+The keyword argument `check` can be set to `false` to skip verification whether `gcm` is indeed a generalized Cartan matrix.
+
+# Example
+```jldoctest
+julia> cartan_bilinear_form(cartan_matrix(:B, 2))
+[ 4   -2]
+[-2    2]
+"""
+function cartan_bilinear_form(gcm::ZZMatrix; check::Bool=true)
+  @req !check || is_cartan_matrix(gcm) "requires a generalized Cartan matrix"
+
+  bil = deepcopy(gcm)
+  sym = cartan_symmetrizer(gcm; check=false)
+  for i in 1:length(sym)
+    mul!(view(bil, i, :), sym[i])
+  end
+  return bil
+end
+
+@doc raw"""
+    cartan_type(gcm::ZZMatrix; check::Bool=true) -> Vector{Tuple{Symbol, Int}}
+
+Returns the Cartan type of a Cartan matrix `gcm` (currently only Cartan matrices of finite type are supported).
+This function is left inverse to `cartan_matrix`, i.e. in the case of isomorphic types (e.g. $B_2$ and $C_2$)
+the ordering of the roots does matter (see the example below).
+The keyword argument `check` can be set to `false` to skip verification whether `gcm` is indeed a generalized Cartan matrix.
```

Comment:
```suggestion
The keyword argument `check` can be set to `false` to skip verification whether `gcm` is indeed a Cartan matrix of finite type.
```
You state above that this only works for finite type currently.